libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -0,0 +1,204 @@
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/Visitor.h>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+void appendMemoryStore(
+	vector<Statement>& _statements,
+	langutil::SourceLocation const& _loc,
+	YulString _mpos,
+	Expression _value
+)
+{
+	_statements.emplace_back(ExpressionStatement{_loc, FunctionCall{
+		_loc,
+		Identifier{_loc, "mstore"_yulstring},
+		{
+			Literal{_loc, LiteralKind::Number, _mpos, {}},
+			std::move(_value)
+		}
+	}});
+}
+}
+
+StackToMemoryMover::StackToMemoryMover(
+	OptimiserStepContext& _context,
+	u256 _reservedMemory,
+	map<YulString, map<YulString, uint64_t>> const& _memorySlots
+): m_reservedMemory(std::move(_reservedMemory)), m_memorySlots(_memorySlots), m_nameDispenser(_context.dispenser)
+{
+	auto const* evmDialect = dynamic_cast<EVMDialect const*>(&_context.dialect);
+	yulAssert(
+		evmDialect && evmDialect->providesObjectAccess(),
+		"StackToMemoryMover can only be run on objects using the EVMDialect with object access."
+	);
+
+	if (m_memorySlots.count(YulString{}))
+		m_currentFunctionMemorySlots = &m_memorySlots.at(YulString{});
+}
+
+void StackToMemoryMover::operator()(FunctionDefinition& _functionDefinition)
+{
+	map<YulString, uint64_t> const* saved = m_currentFunctionMemorySlots;
+	if (m_memorySlots.count(_functionDefinition.name))
+	{
+		m_currentFunctionMemorySlots = &m_memorySlots.at(_functionDefinition.name);
+		for (TypedName const& param: _functionDefinition.parameters + _functionDefinition.returnVariables)
+			if (m_currentFunctionMemorySlots->count(param.name))
+			{
+				// TODO: we cannot handle function parameters yet.
+				m_currentFunctionMemorySlots = nullptr;
+				break;
+			}
+	}
+	else
+		m_currentFunctionMemorySlots = nullptr;
+	ASTModifier::operator()(_functionDefinition);
+	m_currentFunctionMemorySlots = saved;
+}
+
+void StackToMemoryMover::operator()(Block& _block)
+{
+	using OptionalStatements = std::optional<vector<Statement>>;
+	if (!m_currentFunctionMemorySlots)
+	{
+		ASTModifier::operator()(_block);
+		return;
+	}
+	auto containsVariableNeedingEscalation = [&](auto const& _variables) {
+		return util::contains_if(_variables, [&](auto const& var) {
+			return m_currentFunctionMemorySlots->count(var.name);
+		});
+	};
+	auto rewriteAssignmentOrVariableDeclaration = [&](
+		langutil::SourceLocation const& _loc,
+		auto const& _variables,
+		std::unique_ptr<Expression> _value
+	) -> std::vector<Statement> {
+		if (_variables.size() == 1)
+		{
+			std::vector<Statement> result;
+			appendMemoryStore(
+				result,
+				_loc,
+				getMemoryOffset(_variables.front().name),
```

Comment:
I think there are no checks until here that the variable is actually in the map. If it isn't, `memoryOffset` will throw.

Reply:
Note that all this is in a lambda that's only called later in ``util::iterateReplacing``... and the calls there are all guarded by ``if (!containsVariableNeedingEscalation(_varDecl.variables)) return defaultVisit();`` (which makes sure it is in the map)...
But maybe worth thinking about how to make this easier to read... the problem is that the lambdas ``containsVariableNeedingEscalation`` and ``rewriteAssignmentOrVariableDeclaration`` both need access to members and need to be templated - and I don't want to move them to member functions implemented in the header...